File: offset-manager-remote/src/main/java/localoffsetmanager/error/ErrorResponse.java
package localoffsetmanager.error;

import com.fasterxml.jackson.annotation.JsonInclude;

import java.util.Collections;
import java.util.Map;

public record ErrorResponse(
  int statusCode,
  String type,
  String message,
  @JsonInclude(JsonInclude.Include.NON_EMPTY)
  Map<String, String> properties
) {
  public static ErrorResponse of(ErrorType errorType, String message) {
    return new ErrorResponse(
      errorType.getHttpStatusCode(),
      errorType.name(),
      message,
      Collections.emptyMap()
    );
  }

  public static ErrorResponse of(
    ErrorType errorType,
    String message,
    Map<String, String> properties
  ) {
    return new ErrorResponse(
      errorType.getHttpStatusCode(),
      errorType.name(),
      message,
      properties
    );
  }
}

File: offset-manager/src/main/kotlin/offsetmanager/error/ErrorResponse.kt
package offsetmanager.error

import com.fasterxml.jackson.annotation.JsonInclude

data class ErrorResponse(
  val statusCode: Int,
  val type: String,
  val message: String,
  @JsonInclude(JsonInclude.Include.NON_EMPTY)
  val properties: Map<String, String?>
) {
  companion object {
    @JvmStatic
    fun of(errorType: ErrorType, message: String): ErrorResponse {
      return ErrorResponse(
        errorType.httpStatusCode,
        errorType.name,
        message,
        mapOf()
      )
    }

    @JvmStatic
    fun of(
      errorType: ErrorType,
      message: String,
      properties: MutableMap<String, String?>
    ): ErrorResponse {
      return ErrorResponse(
        errorType.httpStatusCode,
        errorType.name,
        message,
        properties
      )
    }
  }
}
